ingester/src/init/wal_replay.rs

```diff
@@ -174,21 +176,54 @@ where
     Ok(max_sequence)
 }
 
-/// Replay the entries in `file`, applying them to `buffer`. Returns the highest
-/// sequence number observed in the file, or [`None`] if the file was empty.
-async fn replay_file<T>(
-    file: wal::ClosedSegmentFileReader,
+// A trait to associate a [`SegmentId`] with a WAL op batch reader
+trait SegmentedWalOpBatchReader:
+    Iterator<Item = Result<Vec<SequencedWalOp>, wal::Error>> + Send
+{
+    fn id(&self) -> SegmentId;
+}
+
+// Implement the trait for the [`wal::ClosedSegmentFileReader`]
+impl SegmentedWalOpBatchReader for wal::ClosedSegmentFileReader {
+    fn id(&self) -> SegmentId {
+        self.id()
+    }
+}
+
+/// Replay the entries in `batches`, applying them to `buffer`. Returns the
+/// highest sequence number observed across the batches, or [`None`] if there
+/// were no entries read.
+///
+/// # Warnings
+///
+/// This function relies on the [`wal::Error::UnableToReadNextOps`] error
+/// meaning that there are no more valid completed writes which can be read
+/// from the provided `batches` and that it is safe to ignore them.
+async fn replay_sequenced_op_batches<T, W>(
+    batches: W,
     sink: &T,
     ok_op_count_metric: &U64Counter,
     empty_op_count_metric: &U64Counter,
 ) -> Result<Option<SequenceNumber>, WalReplayError>
 where
     T: DmlSink,
+    W: SegmentedWalOpBatchReader,
 {
     let mut max_sequence = None;
     let start = Instant::now();
+    let segment_id = batches.id();
+
+    for batch in batches {
+        if let Err(
+            err @ wal::Error::UnableToReadNextOps {
+                source: wal::blocking::ReaderError::IncompleteEntry { .. },
+            },
+        ) = batch
+        {
+            error!(%err, ?segment_id, "unable to recover further op batches from wal segment");
```

Comment:
This sounds like something unexpected has happened that is fatal, when in fact we're handling it:

```suggestion
            warn!(%err, ?segment_id, "detected truncated WAL write, skipping replay for file");
```